src/orquestra/quantum/wip/operators/_pauli_operators.py

```diff
@@ -0,0 +1,546 @@
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import copy
+import re
+import warnings
+from collections import OrderedDict
+from itertools import chain, product
+from typing import (
+    Any,
+    Dict,
+    FrozenSet,
+    Hashable,
+    Iterator,
+    List,
+    Sequence,
+    Set,
+    Tuple,
+    Union,
+    cast,
+)
+
+import numpy as np
+
+from orquestra.quantum.circuits import Circuit, Operation, builtin_gate_by_name
+
+CoefficientTypes = Union[int, float, complex]
+PauliRepresentation = Union["PauliTerm", "PauliSum"]
+
+allowed_operators = ["X", "Y", "Z", "I"]
+
+# Use ASCII values instead of actual strings
+# as keys to allow for commutativity
+OPERATOR_MAP = {
+    ord("X") + ord("Y"): "Z",
+    ord("Y") + ord("Z"): "X",
+    ord("X") + ord("Z"): "Y",
+}
+
+COEFF_MAP = {
+    "XY": 1.0j,
+    "XZ": -1.0j,
+    "YX": -1.0j,
+    "YZ": 1.0j,
+    "ZX": 1.0j,
+    "ZY": -1.0j,
+}
+
+HASH_PRECISION = 1e6
+
+
+def _efficient_exponentiation(
+    pauli_rep: PauliRepresentation, power: int
+) -> PauliRepresentation:
+    """
+    A more efficient implementation of exponentiation.
+    Assumes validation checks already done from parent routine.
+    """
+    if power == 0:
+        return type(pauli_rep).identity()
+
+    if power % 2 == 1:
+        return pauli_rep * _efficient_exponentiation(pauli_rep, power - 1)
+
+    intermediate_result = _efficient_exponentiation(pauli_rep, power // 2)
+    return intermediate_result * intermediate_result
+
+
+def _validate_type(object: Any) -> None:
+    if not isinstance(object, (PauliSum, PauliTerm, int, float, complex)):
+        raise TypeError(
+            f"Can't carry out operation with object of type {type(object)}."
+        )
+
+
+class PauliTerm:
+    """
+    A datastructure for storing information about a single Pauli Term
+    """
+
+    def __init__(
+        self,
+        operator: Union[Tuple[str, int], str],
+        coefficient: CoefficientTypes = 1.0,
+    ):
+        if isinstance(operator, str):
+            matched_pattern = re.match(r"([a-zA-Z]+)([0-9]+)", operator, re.I)
+
+            if not matched_pattern:
+                raise ValueError("Badly formatted string representation passed.")
+
+            op, qubit_idx = matched_pattern.groups()
+        elif isinstance(operator, tuple):
+            op, qubit_idx = operator
+
+        qubit_idx = int(qubit_idx)
+
+        if op not in allowed_operators:
+            raise ValueError(f"Passed operator not supported. Got {op}.")
+
+        self._ops: Dict[int, str] = OrderedDict()
```

Comment:
I am probably missing something, but why are we specifically using `OrderedDict` here?